File: ios-native/Sources/Encatch/Core/EncatchStorage.swift
import Foundation

/// Key-value persistence layer, mirroring `storage.ts` from the RN SDK. All keys are
/// namespaced under `@encatch/`. Backed by `UserDefaults` — this matches the Kotlin source's use
/// of `Settings`/SharedPreferences, which is also non-secure storage, so `UserDefaults` is the
/// correct parity choice (not Keychain).
public final class EncatchStorage: @unchecked Sendable {
    private static let keyDeviceId = "@encatch/device_id"
    private static let keyUserName = "@encatch/user_name"
    private static let keyUserIdPrefix = "@encatch/user_id_"
    private static let keyFtPrefix = "@encatch/ft_"
    private static let keyPreferences = "@encatch/preferences"
    private static let keySessionStopped = "@encatch/session_stopped"
    private static let keyRetryQueue = "@encatch/retry_queue"

    private let defaults: UserDefaults
    private let lock = NSLock()

    // Session ID is in-memory only — reset when the app process ends.
    private var inMemorySessionId: String?

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    public func getOrCreateDeviceId() -> String {
        if let stored = defaults.string(forKey: Self.keyDeviceId) {
            return stored
        }
        let id = uuidV7()
        defaults.set(id, forKey: Self.keyDeviceId)
        return id
    }

    public func getOrCreateSessionId() -> String {
        lock.lock()
        defer { lock.unlock() }
        if let inMemorySessionId {
            return inMemorySessionId
        }
        let id = uuidV7()
        inMemorySessionId = id
        return id
    }

    public func clearSession() {
        lock.lock()
        inMemorySessionId = nil
        lock.unlock()
    }

    public func getUserName() -> String? {
        defaults.string(forKey: Self.keyUserName)
    }

    public func setUserName(_ name: String) {
        defaults.set(name, forKey: Self.keyUserName)
    }

    public func clearUserName() {
        defaults.removeObject(forKey: Self.keyUserName)
    }

    public func getUserId(userName: String) -> String? {
        defaults.string(forKey: Self.keyUserIdPrefix + userName)
    }

    public func setUserId(userName: String, userId: String) {
        defaults.set(userId, forKey: Self.keyUserIdPrefix + userName)
    }

    public func clearUserId(userName: String) {
        defaults.removeObject(forKey: Self.keyUserIdPrefix + userName)
    }

    private func ftKey(_ identityKey: String) -> String { Self.keyFtPrefix + identityKey }

    public func getFeedbackTransactions(identityKey: String) -> String? {
        defaults.string(forKey: ftKey(identityKey))
    }

    public func setFeedbackTransactions(identityKey: String, value: String) {
        defaults.set(value, forKey: ftKey(identityKey))
    }

    public func clearFeedbackTransactions(identityKey: String) {
        defaults.removeObject(forKey: ftKey(identityKey))
    }

    public struct Preferences: Sendable {
        public var locale: String?
        public var country: String?

        public init(locale: String? = nil, country: String? = nil) {
            self.locale = locale
            self.country = country
        }
    }

    public func getPreferences() -> Preferences {
        guard let raw = defaults.string(forKey: Self.keyPreferences),
              let data = raw.data(using: .utf8),
              let map = try? JSONDecoder().decode([String: String].self, from: data) else {
            return Preferences()
        }
        return Preferences(locale: map["locale"], country: map["country"])
    }

    public func setPreferences(locale: String? = nil, country: String? = nil) {
        let current = getPreferences()
        let merged = Preferences(locale: locale ?? current.locale, country: country ?? current.country)
        var map: [String: String] = [:]
        if let locale = merged.locale { map["locale"] = locale }
        if let country = merged.country { map["country"] = country }
        guard let data = try? JSONEncoder().encode(map), let json = String(data: data, encoding: .utf8) else {
            return
        }
        defaults.set(json, forKey: Self.keyPreferences)
    }

    public func clearPreferences() {
        defaults.removeObject(forKey: Self.keyPreferences)
    }

    public func getSessionStopped() -> Bool {
        defaults.string(forKey: Self.keySessionStopped) == "true"
    }

    public func setSessionStopped() {
        defaults.set("true", forKey: Self.keySessionStopped)
    }

    public func clearSessionStopped() {
        defaults.removeObject(forKey: Self.keySessionStopped)
    }

    public func getRetryQueueRaw() -> String? {
        defaults.string(forKey: Self.keyRetryQueue)
    }

    public func setRetryQueueRaw(_ json: String) {
        defaults.set(json, forKey: Self.keyRetryQueue)
    }

    public func clearRetryQueue() {
        defaults.removeObject(forKey: Self.keyRetryQueue)
    }

    /// Wipes every `@encatch/`-namespaced key — used by `Encatch.clearAll`.
    public func clearAll() {
        defaults.removeObject(forKey: Self.keyDeviceId)
        defaults.removeObject(forKey: Self.keyUserName)
        defaults.removeObject(forKey: Self.keyPreferences)
        defaults.removeObject(forKey: Self.keySessionStopped)
        defaults.removeObject(forKey: Self.keyRetryQueue)
        for key in defaults.dictionaryRepresentation().keys
        where key.hasPrefix(Self.keyUserIdPrefix) || key.hasPrefix(Self.keyFtPrefix) {
            defaults.removeObject(forKey: key)
        }
        lock.lock()
        inMemorySessionId = nil
        lock.unlock()
    }
}

File: core/src/commonMain/kotlin/com/encatch/core/EncatchStorage.kt
package com.encatch.core

import com.russhwolf.settings.Settings
import com.russhwolf.settings.set
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Key-value persistence layer, mirroring `storage.ts` from the RN SDK. All keys are
 * namespaced under `@encatch/`. Backed by [Settings] (SharedPreferences on Android).
 */
class EncatchStorage internal constructor(
    private val settings: Settings = createEncatchSettings(),
) {
    companion object {
        private const val KEY_DEVICE_ID = "@encatch/device_id"
        private const val KEY_USER_NAME = "@encatch/user_name"
        private const val KEY_USER_ID_PREFIX = "@encatch/user_id_"
        private const val KEY_FT_PREFIX = "@encatch/ft_"
        private const val KEY_PREFERENCES = "@encatch/preferences"
        private const val KEY_SESSION_STOPPED = "@encatch/session_stopped"
        private const val KEY_RETRY_QUEUE = "@encatch/retry_queue"
    }

    // Session ID is in-memory only — reset when the app process ends.
    private var inMemorySessionId: String? = null

    fun getOrCreateDeviceId(): String {
        val stored = runCatching { settings.getStringOrNull(KEY_DEVICE_ID) }.getOrNull()
        if (stored != null) return stored
        val id = uuidV7()
        runCatching { settings[KEY_DEVICE_ID] = id }
        return id
    }

    fun getOrCreateSessionId(): String {
        inMemorySessionId?.let { return it }
        val id = uuidV7()
        inMemorySessionId = id
        return id
    }

    fun clearSession() {
        inMemorySessionId = null
    }

    fun getUserName(): String? = runCatching { settings.getStringOrNull(KEY_USER_NAME) }.getOrNull()

    fun setUserName(name: String) {
        runCatching { settings[KEY_USER_NAME] = name }
    }

    fun clearUserName() {
        runCatching { settings.remove(KEY_USER_NAME) }
    }

    fun getUserId(userName: String): String? =
        runCatching { settings.getStringOrNull(KEY_USER_ID_PREFIX + userName) }.getOrNull()

    fun setUserId(userName: String, userId: String) {
        runCatching { settings[KEY_USER_ID_PREFIX + userName] = userId }
    }

    fun clearUserId(userName: String) {
        runCatching { settings.remove(KEY_USER_ID_PREFIX + userName) }
    }

    private fun ftKey(identityKey: String) = KEY_FT_PREFIX + identityKey

    fun getFeedbackTransactions(identityKey: String): String? =
        runCatching { settings.getStringOrNull(ftKey(identityKey)) }.getOrNull()

    fun setFeedbackTransactions(identityKey: String, value: String) {
        runCatching { settings[ftKey(identityKey)] = value }
    }

    fun clearFeedbackTransactions(identityKey: String) {
        runCatching { settings.remove(ftKey(identityKey)) }
    }

    data class Preferences(val locale: String? = null, val country: String? = null)

    fun getPreferences(): Preferences {
        val raw = runCatching { settings.getStringOrNull(KEY_PREFERENCES) }.getOrNull() ?: return Preferences()
        return runCatching {
            val map = Json.decodeFromString<Map<String, String>>(raw)
            Preferences(locale = map["locale"], country = map["country"])
        }.getOrDefault(Preferences())
    }

    fun setPreferences(locale: String? = null, country: String? = null) {
        val current = getPreferences()
        val merged = Preferences(
            locale = locale ?: current.locale,
            country = country ?: current.country,
        )
        val map = buildMap {
            merged.locale?.let { put("locale", it) }
            merged.country?.let { put("country", it) }
        }
        runCatching { settings[KEY_PREFERENCES] = Json.encodeToString(map) }
    }

    fun clearPreferences() {
        runCatching { settings.remove(KEY_PREFERENCES) }
    }

    fun getSessionStopped(): Boolean =
        runCatching { settings.getStringOrNull(KEY_SESSION_STOPPED) }.getOrNull() == "true"

    fun setSessionStopped() {
        runCatching { settings[KEY_SESSION_STOPPED] = "true" }
    }

    fun clearSessionStopped() {
        runCatching { settings.remove(KEY_SESSION_STOPPED) }
    }

    fun getRetryQueueRaw(): String? = runCatching { settings.getStringOrNull(KEY_RETRY_QUEUE) }.getOrNull()

    fun setRetryQueueRaw(json: String) {
        runCatching { settings[KEY_RETRY_QUEUE] = json }
    }

    fun clearRetryQueue() {
        runCatching { settings.remove(KEY_RETRY_QUEUE) }
    }

    /** Wipes every `@encatch/`-namespaced key — used by [Encatch.clearAll]. */
    fun clearAll() {
        runCatching {
            settings.remove(KEY_DEVICE_ID)
            settings.remove(KEY_USER_NAME)
            settings.remove(KEY_PREFERENCES)
            settings.remove(KEY_SESSION_STOPPED)
            settings.remove(KEY_RETRY_QUEUE)
            settings.keys.filter { it.startsWith(KEY_USER_ID_PREFIX) || it.startsWith(KEY_FT_PREFIX) }
                .forEach { settings.remove(it) }
        }
        inMemorySessionId = null
    }
}
